frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseDiscoverAction.java

```diff
@@ -0,0 +1,78 @@
+import io.vavr.control.Try;
+import lombok.CustomLog;
+
+import com.virtuslab.binding.RuntimeBinding;
+import com.virtuslab.branchlayout.api.readwrite.IBranchLayoutWriter;
+import com.virtuslab.gitmachete.backend.api.IGitMacheteRepositoryCache;
+import com.virtuslab.gitmachete.backend.api.IGitMacheteRepositorySnapshot;
+import com.virtuslab.gitmachete.frontend.actions.dialogs.GraphTableDialog;
+import com.virtuslab.gitmachete.frontend.actions.expectedkeys.IExpectsKeyProject;
+import com.virtuslab.gitmachete.frontend.ui.api.table.BaseGraphTable;
+import com.virtuslab.gitmachete.frontend.ui.providerservice.SelectedGitRepositoryProvider;
+import com.virtuslab.gitmachete.frontend.vfsutils.GitVfsUtils;
+import com.virtuslab.logger.IEnhancedLambdaLogger;
+
+@CustomLog
+public class BaseDiscoverAction extends DumbAwareAction implements IExpectsKeyProject {
+  @Override
+  public void actionPerformed(AnActionEvent anActionEvent) {
+    var project = getProject(anActionEvent);
+    var selectedRepoProvider = project.getService(SelectedGitRepositoryProvider.class).getGitRepositorySelectionProvider();
+    var selectedRepositoryOption = selectedRepoProvider.getSelectedGitRepository();
+    assert selectedRepositoryOption.isDefined() : "Selected repository is undefined";
+    var selectedRepository = selectedRepositoryOption.get();
+    var mainDirPath = GitVfsUtils.getMainDirectoryPath(selectedRepository).toAbsolutePath();
+    var gitDirPath = GitVfsUtils.getGitDirectoryPath(selectedRepository).toAbsolutePath();
+    Try.of(() -> RuntimeBinding.instantiateSoleImplementingClass(IGitMacheteRepositoryCache.class)
+        .getInstance(mainDirPath, gitDirPath).discoverLayoutAndCreateSnapshot())
+        .onFailure(e -> GuiUtils.invokeLaterIfNeeded(() -> VcsNotifier.getInstance(project)
+            .notifyError(/* title */ getString("action.GitMachete.BaseDiscoverAction.repository-discover-error-title"),
+                /* message */ e.getMessage() != null ? e.getMessage() : ""),
+            NON_MODAL))
+        .onSuccess(repoSnapshot -> GuiUtils.invokeLaterIfNeeded(() -> GraphTableDialog.of(repoSnapshot,
+            /* windowTitle */ getString("action.GitMachete.BaseDiscoverAction.discovered-branch-tree-dialog.title"),
+            /* okAction */ repositorySnapshot -> saveDiscoveredLayout(repositorySnapshot,
+                GitVfsUtils.getMacheteFilePath(selectedRepository), project, getGraphTable(anActionEvent)),
+            /* okButtonText */ getString("action.GitMachete.BaseDiscoverAction.discovered-branch-tree-dialog.save-button-text"),
+            /* cancelButtonVisible */ true).show(), NON_MODAL));
+  }
+
+  private void saveDiscoveredLayout(IGitMacheteRepositorySnapshot repositorySnapshot, Path macheteFilePath, Project project,
+      BaseGraphTable gitMacheteGraphTable) {
+    var branchLayoutOption = repositorySnapshot.getBranchLayout();
+    assert branchLayoutOption.isDefined() : "Branch layout from discovered repository is not defined";
```

Comment:
Again, rather break the flow and communicate the error in case branch layout is missing; do not assume it's present

Reply:
same as with selected repository